core/ordersync_subprotocols.go

```diff
@@ -0,0 +1,129 @@
+type FilteredPaginationSubProtocol struct {
+	app     *App
+	perPage int
+}
+
+// NewFilteredPaginationSubprotocol creates and returns a new FilteredPaginationSubprotocol
+// which will respond with perPage orders for each individual request/response.
+func NewFilteredPaginationSubprotocol(app *App, perPage int) *FilteredPaginationSubProtocol {
+	return &FilteredPaginationSubProtocol{
+		app:     app,
+		perPage: perPage,
+	}
+}
+
+// FilteredPaginationRequestMetadata is the request metadata for the
+// FilteredPaginationSubProtocol. It keeps track of the current page and SnapshotID,
+// which is expected to be an empty string on the first request.
+type FilteredPaginationRequestMetadata struct {
+	Page       int    `json:"page"`
+	SnapshotID string `json:"snapshotID"`
+}
+
+// FilteredPaginationResponseMetadata is the response metadata for the
+// FilteredPaginationSubProtocol. It keeps track of the current page and SnapshotID.
+type FilteredPaginationResponseMetadata struct {
+	Page       int    `json:"page"`
+	SnapshotID string `json:"snapshotID"`
+}
+
+// Name reutrns the name of the FilteredPaginationSubProtocol
+func (p *FilteredPaginationSubProtocol) Name() string {
+	return "/pagination-with-filter/version/0"
+}
+
+// GetOrders returns the orders for one page, based on the page number
+// and snapshotID in the corresponding to the given request. This is
+// the implementation for the "provider" side of the subprotocol.
+func (p *FilteredPaginationSubProtocol) GetOrders(ctx context.Context, req *ordersync.Request) (*ordersync.Response, error) {
+	var metadata *FilteredPaginationRequestMetadata
+	if req.Metadata == nil {
+		// Default metadata for the first request.
+		metadata = &FilteredPaginationRequestMetadata{
+			Page:       0,
+			SnapshotID: "",
+		}
+	} else {
+		var ok bool
+		metadata, ok = req.Metadata.(*FilteredPaginationRequestMetadata)
+		if !ok {
+			return nil, fmt.Errorf("FilteredPaginationSubProtocol received request with wrong metadata type (got %T)", req.Metadata)
+		}
+	}
+
+	ordersResp, err := p.app.GetOrders(metadata.Page, p.perPage, metadata.SnapshotID)
+	if err != nil {
+		return nil, err
+	}
+	orders := make([]*zeroex.SignedOrder, len(ordersResp.OrdersInfos))
+	for i, orderInfo := range ordersResp.OrdersInfos {
+		orders[i] = orderInfo.SignedOrder
+	}
+	// TODO(albrow): Filter orders
+	return &ordersync.Response{
+		Orders:   orders,
+		Complete: len(orders) == 0,
+		Metadata: &FilteredPaginationResponseMetadata{
+			Page:       metadata.Page,
+			SnapshotID: ordersResp.SnapshotID,
+		},
+	}, nil
+}
+
+// HandleOrders handles the orders for one page by validating them, storing them in the database,
+// and firing the appropriate events. It also returns the next request to be sent. This is the
+// implementation for the "requester" side of the subprotocol.
+func (p *FilteredPaginationSubProtocol) HandleOrders(ctx context.Context, res *ordersync.Response) (*ordersync.Request, error) {
```

Comment:
Same as above, consider switching the name of this. Maybe names like `HandleOrderSyncRequest` and `HandleOrderSyncResponse`?

Reply:
I like this suggestion and just applied it. Would prefer to keep `HandleStream` the same since it is a carryover from libp2p semantics. `ordersync.Service` is a stream handler, and other stream handlers use this naming convention in libp2p.